ports/unix/modffi.c

```diff
@@ -266,19 +269,51 @@ STATIC mp_obj_t mod_ffi_func(mp_obj_t rettype, mp_obj_t addr_in, mp_obj_t argtyp
 }
 MP_DEFINE_CONST_FUN_OBJ_3(mod_ffi_func_obj, mod_ffi_func);
 
-STATIC void call_py_func(ffi_cif *cif, void *ret, void **args, void *func) {
+STATIC void call_py_func(ffi_cif *cif, void *ret, void **args, void *user_data) {
     mp_obj_t pyargs[cif->nargs];
+    mp_obj_fficallback_t *o = user_data;
+    mp_obj_t pyfunc = o->pyfunc;
+
+    if (o->lock) {
+        #if MICROPY_ENABLE_SCHEDULER
+        mp_sched_lock();
+        #endif
+        gc_lock();
+    }
+
     for (uint i = 0; i < cif->nargs; i++) {
         pyargs[i] = mp_obj_new_int(*(mp_int_t *)args[i]);
     }
-    mp_obj_t res = mp_call_function_n_kw(MP_OBJ_FROM_PTR(func), cif->nargs, 0, pyargs);
+    mp_obj_t res = mp_call_function_n_kw(pyfunc, cif->nargs, 0, pyargs);
```

Comment:
If this call raises an exception, and locking is enabled, then things will not go well, the GC will remain locked and the code will fail completely.

I think it makes sense to have two separate callback functions: `call_py_func` and `call_py_func_with_lock`.  The correct one is registered with `ffi_prep_closure_loc`.  The wrapper `call_py_func_with_lock` will do something similar to `lib/utils/mpirq.c:mp_irq_handler`.

Reply:
Done.